File: src/main/java/com/projectswg/holocore/services/gameplay/combat/loot/GrantLootService.java
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.services.gameplay.combat.loot;

import com.projectswg.common.data.encodables.oob.ProsePackage;
import com.projectswg.common.data.encodables.oob.StringId;
import com.projectswg.common.data.location.Location;
import com.projectswg.common.network.packets.swg.zone.ClientOpenContainerMessage;
import com.projectswg.common.network.packets.swg.zone.PlayMusicMessage;
import com.projectswg.common.network.packets.swg.zone.StopClientEffectObjectByLabelMessage;
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupCloseLotteryWindow;
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupOpenLotteryWindow;
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupRequestLotteryItems;
import com.projectswg.holocore.intents.gameplay.combat.CreatureKilledIntent;
import com.projectswg.holocore.intents.gameplay.combat.loot.CorpseLootedIntent;
import com.projectswg.holocore.intents.gameplay.combat.loot.LootItemIntent;
import com.projectswg.holocore.intents.gameplay.combat.loot.LootLotteryStartedIntent;
import com.projectswg.holocore.intents.gameplay.combat.loot.LootRequestIntent;
import com.projectswg.holocore.intents.gameplay.combat.loot.LootRequestIntent.LootType;
import com.projectswg.holocore.intents.support.global.chat.SystemMessageIntent;
import com.projectswg.holocore.intents.support.global.network.InboundPacketIntent;
import com.projectswg.holocore.intents.support.objects.swg.ContainerTransferIntent;
import com.projectswg.holocore.intents.support.objects.swg.DestroyObjectIntent;
import com.projectswg.holocore.resources.support.data.config.ConfigFile;
import com.projectswg.holocore.resources.support.data.server_info.DataManager;
import com.projectswg.holocore.resources.support.global.player.Player;
import com.projectswg.holocore.resources.support.objects.permissions.ContainerPermissionsType;
import com.projectswg.holocore.resources.support.objects.swg.SWGObject;
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject;
import com.projectswg.holocore.resources.support.objects.swg.custom.AIObject;
import com.projectswg.holocore.resources.support.objects.swg.group.GroupObject;
import com.projectswg.holocore.resources.support.objects.swg.group.LootRule;
import com.projectswg.holocore.resources.support.objects.swg.tangible.CreditObject;
import com.projectswg.holocore.services.support.objects.ObjectStorageService.ObjectLookup;
import me.joshlarson.jlcommon.concurrency.ScheduledThreadPool;
import me.joshlarson.jlcommon.control.IntentHandler;
import me.joshlarson.jlcommon.control.Service;
import me.joshlarson.jlcommon.utilities.Arguments;

import java.util.*;
import java.util.Map.Entry;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;

public final class GrantLootService extends Service {
	
	private final Map<CreatureObject, CorpseLootRestrictions> lootRestrictions;
	private final ScheduledThreadPool executor;
	private final int lootRange;
	
	public GrantLootService() {
		this.lootRestrictions = new ConcurrentHashMap<>();
		this.executor = new ScheduledThreadPool(1, "grant-loot-service");
		this.lootRange = DataManager.getConfig(ConfigFile.LOOTOPTIONS).getInt("LOOT-RANGE", 64);
	}
	
	@Override
	public boolean start() {
		executor.start();
		return true;
	}
	
	@Override
	public boolean stop() {
		executor.stop();
		return executor.awaitTermination(1000);
	}
	
	@IntentHandler
	private void handleCreatureKilledIntent(CreatureKilledIntent cki) {
		CreatureObject corpse = cki.getCorpse();
		if (corpse.isPlayer())
			return;
		
		Location corpseWorldLocation = corpse.getWorldLocation();
		CreatureObject highestDamageDealer = corpse.getHighestDamageDealer();
		if (highestDamageDealer == null || !highestDamageDealer.isPlayer())
			return;
		long groupId = highestDamageDealer.getGroupId();
		
		List<CreatureObject> looters = new ArrayList<>();
		if (groupId != 0) {
			GroupObject group = (GroupObject) ObjectLookup.getObjectById(groupId);
			assert group != null;
			switch (group.getLootRule()) {
				case MASTER_LOOTER:
					looters.add((CreatureObject) ObjectLookup.getObjectById(group.getLootMaster()));
					break;
				case FREE_FOR_ALL:
				case RANDOM:
				case LOTTERY:
					// Get all looters within range
					looters.addAll(group.getGroupMemberObjects().stream().filter(m -> m.getWorldLocation().flatDistanceTo(corpseWorldLocation) <= lootRange).collect(Collectors.toList()));
					break;
			}
			if (group.getLootRule() == LootRule.LOTTERY) {
				lootRestrictions.put(corpse, new LotteryLootRestrictions(corpse, looters));
				return;
			} else if (group.getLootRule() == LootRule.RANDOM) {
				lootRestrictions.put(corpse, new RandomLootRestrictions(corpse, looters));
				return;
			}
		} else {
			looters.add(highestDamageDealer);
		}
		lootRestrictions.put(corpse, new StandardLootRestrictions(corpse, looters));
	}
	
	@IntentHandler
	private void handleDestroyObjectIntent(DestroyObjectIntent doi) {
		// For when the corpse is gone
		SWGObject object = doi.getObject();
		if (object instanceof CreatureObject) {
			CorpseLootRestrictions restrictions = lootRestrictions.remove(object);
			if (restrictions != null)
				restrictions.setValid(false);
		}
	}
	
	@IntentHandler
	private void handleInboundPacketIntent(InboundPacketIntent ipi) {
		if (!(ipi.getPacket() instanceof GroupRequestLotteryItems))
			return;
		Player player = ipi.getPlayer();
		GroupRequestLotteryItems request = (GroupRequestLotteryItems) ipi.getPacket();
		SWGObject requestObject = ObjectLookup.getObjectById(request.getInventoryId());
		if (requestObject == null)
			return;
		requestObject = requestObject.getParent();
		if (!(requestObject instanceof CreatureObject))
			return;
		CorpseLootRestrictions restriction = lootRestrictions.get(requestObject);
		if (!(restriction instanceof LotteryLootRestrictions))
			return;
		((LotteryLootRestrictions) restriction).updatePreferences(player.getCreatureObject(), request.getRequestedItems().stream().map(ObjectLookup::getObjectById).filter(Objects::nonNull).collect(Collectors.toList()));
	}
	
	@IntentHandler
	private void handleContainerTransfer(ContainerTransferIntent cti) {
		// Only check transfers into players
		SWGObject container = cti.getContainer();
		if (container == null || container.getOwner() == null)
			return;
		
		// If this is a looted item, reset the container permissions to default
		SWGObject object = cti.getObject();
		if (object.getContainerPermissions() == ContainerPermissionsType.LOOT)
			object.setContainerPermissions(ContainerPermissionsType.DEFAULT);
	}
	
	@IntentHandler
	private void handleLootRequestIntent(LootRequestIntent lri) {
		Player player = lri.getPlayer();
		CreatureObject looter = player.getCreatureObject();
		CreatureObject corpse = lri.getTarget();
		
		Arguments.validate(corpse instanceof AIObject, "Attempted to loot a non-AI object");
		CorpseLootRestrictions restrictions = lootRestrictions.get(corpse);
		if (restrictions == null || !restrictions.canLoot(looter)) {
			SystemMessageIntent.broadcastPersonal(player, "You don't have permission to loot '"+corpse.getObjectName()+ '\'');
			return;
		}
		
		if (restrictions instanceof LotteryLootRestrictions) {
			if (((LotteryLootRestrictions) restrictions).isStarted())
				return; // Don't keep re-starting the lottery
			executor.execute(30000, () -> {
				// Close the windows after 30 seconds, and grant the items won
				if (!restrictions.isValid())
					return;
				((LotteryLootRestrictions) restrictions).commitLottery();
			});
		}
		restrictions.handle(looter, lri.getType());
	}
	
	@IntentHandler
	private void handleLootItemIntent(LootItemIntent lii) {
		CreatureObject looter = lii.getLooter();
		CreatureObject corpse = lii.getCorpse();
		SWGObject item = lii.getItem();
		
		Arguments.validate(corpse instanceof AIObject, "Attempted to loot a non-AI object");
		CorpseLootRestrictions restrictions = lootRestrictions.get(corpse);
		if (restrictions == null || !restrictions.canLoot(looter)) {
			SystemMessageIntent.broadcastPersonal(looter.getOwner(), "You don't have permission to loot '"+corpse.getObjectName()+ '\'');
			return;
		}
		
		restrictions.loot(looter, item);
	}
	
	private static abstract class CorpseLootRestrictions {
		
		private final CreatureObject corpse;
		private final List<CreatureObject> looters;
		private final AtomicBoolean valid;
		
		public CorpseLootRestrictions(CreatureObject corpse, List<CreatureObject> looters) {
			this.corpse = corpse;
			this.looters = looters;
			this.valid = new AtomicBoolean(true);
		}
		
		public boolean isValid() {
			return valid.get();
		}
		
		public CreatureObject getCorpse() {
			return corpse;
		}
		
		public List<CreatureObject> getLooters() {
			return Collections.unmodifiableList(looters);
		}
		
		public boolean canLoot(CreatureObject looter) {
			return looters.contains(looter);
		}
		
		public void setValid(boolean valid) {
			this.valid.set(valid);
		}
		
		public synchronized void loot(CreatureObject looter, SWGObject item) {
			throw new UnsupportedOperationException("this loot restriction cannot do manual looting");
		}
		
		public abstract void handle(CreatureObject looter, LootType type);
		
		protected synchronized void transferItem(CreatureObject looter, SWGObject item) {
			if (!getCorpse().getInventory().getContainedObjects().contains(item) || !canLoot(looter)) {
				return;
			}
			Player player = looter.getOwner();
			
			switch (item.moveToContainer(looter, looter.getSlottedObject("inventory"))) {
				case SUCCESS: {
					String itemName = item.getObjectName();
					
					if (item instanceof CreditObject) {
						onLootedCredits(looter, ((CreditObject) item).getAmount());
					} else {
						new SystemMessageIntent(player, new ProsePackage("StringId", new StringId("loot_n", "solo_looted"), "TO", itemName)).broadcast();
					}
					onLooted(looter, getCorpse());
					break;
				}
				case CONTAINER_FULL:
					new SystemMessageIntent(player, "@container_error_message:container03").broadcast();
					player.sendPacket(new PlayMusicMessage(0, "sound/ui_danger_message.snd", 1, false));
					break;
				case NO_PERMISSION:
					new SystemMessageIntent(player, "@container_error_message:container08").broadcast();
					player.sendPacket(new PlayMusicMessage(0, "sound/ui_negative.snd", 1, false));
					break;
				case SLOT_NO_EXIST:
					new SystemMessageIntent(player, "@container_error_message:container06").broadcast();
					player.sendPacket(new PlayMusicMessage(0, "sound/ui_negative.snd", 1, false));
					break;
				case SLOT_OCCUPIED:
					new SystemMessageIntent(player, "@container_error_message:container08").broadcast();
					player.sendPacket(new PlayMusicMessage(0, "sound/ui_negative.snd", 1, false));
					break;
			}
		}
		
		protected void splitCredits(CreditObject target, Collection<CreatureObject> looters) {
			long amount = target.getAmount() / looters.size();
			assert amount > 0;
			
			for (CreatureObject looter : looters) {
				looter.addToBank(amount);
				onLootedCredits(looter, amount);
			}
			
			DestroyObjectIntent.broadcast(target);
		}
		
		protected void onLooted(CreatureObject looter, CreatureObject corpse) {
			SWGObject lootInventory = corpse.getInventory();
			if (lootInventory.getContainedObjects().isEmpty()) {
				new CorpseLootedIntent(corpse).broadcast();
				
				long looterGroupId = looter.getGroupId();
				if (looterGroupId != 0) {
					GroupObject killerGroup = (GroupObject) ObjectLookup.getObjectById(looterGroupId);
					assert killerGroup != null;
					for (CreatureObject groupMember : killerGroup.getGroupMemberObjects())
						groupMember.getOwner().sendPacket(new StopClientEffectObjectByLabelMessage(corpse.getObjectId(), "lootMe", false));
				} else {
					looter.getOwner().sendPacket(new StopClientEffectObjectByLabelMessage(corpse.getObjectId(), "lootMe", false));
				}
			}
		}
		
		protected void onLootedCredits(CreatureObject looter, long amount) {
			// Perhaps "prose_coin_loot_no_target" is the proper string?
			new SystemMessageIntent(looter.getOwner(), new ProsePackage("StringId", new StringId("base_player", "prose_transfer_success"), "DI", (int) amount)).broadcast();
		}
		
	}
	
	private static class StandardLootRestrictions extends CorpseLootRestrictions {
		
		public StandardLootRestrictions(CreatureObject corpse, List<CreatureObject> looters) {
			super(corpse, looters);
		}
		
		@Override
		public synchronized void loot(CreatureObject looter, SWGObject item) {
			transferItem(looter, item);
		}
		
		@Override
		public synchronized void handle(CreatureObject looter, LootType type) {
			SWGObject lootInventory = getCorpse().getInventory();
			Collection<SWGObject> lootItems = lootInventory.getContainedObjects();
			switch (type) {
				case LOOT: // Open loot box
					if (!lootItems.isEmpty())
						looter.getOwner().sendPacket(new ClientOpenContainerMessage(lootInventory.getObjectId(), ""));
					break;
				case LOOT_ALL: // Request to loot all items
					for (SWGObject loot : lootItems)
						LootItemIntent.broadcast(looter, getCorpse(), loot);
					break;
			}
		}
	}
	
	private static class RandomLootRestrictions extends CorpseLootRestrictions {
		
		public RandomLootRestrictions(CreatureObject corpse, List<CreatureObject> looters) {
			super(corpse, looters);
		}
		
		@Override
		public synchronized void loot(CreatureObject looter, SWGObject item) {
			if (item instanceof CreditObject && ((CreditObject) item).getAmount() >= getLooters().size())
				splitCredits((CreditObject) item, getLooters());
			else
				transferItem(looter, item);
		}
		
		@Override
		public synchronized void handle(CreatureObject looter, LootType type) {
			Collection<SWGObject> lootItems = getCorpse().getInventory().getContainedObjects();
			for (SWGObject loot : lootItems)
				LootItemIntent.broadcast(looter, getCorpse(), loot);
		}
		
	}
	
	private static class LotteryLootRestrictions extends CorpseLootRestrictions {
		
		private final Map<CreatureObject, List<SWGObject>> preferences;
		private final AtomicBoolean started;
		private final AtomicBoolean committed;
		
		public LotteryLootRestrictions(CreatureObject corpse, List<CreatureObject> looters) {
			super(corpse, looters);
			this.preferences = new HashMap<>();
			this.started = new AtomicBoolean(false);
			this.committed = new AtomicBoolean(false);
		}
		
		public boolean isStarted() {
			return started.get();
		}
		
		@Override
		public synchronized void handle(CreatureObject looter, LootType type) {
			if (started.getAndSet(true))
				return;
			if (getCorpse().getInventory().getContainedObjects().isEmpty())
				return;
			LootLotteryStartedIntent.broadcast(getCorpse());
			for (CreatureObject creature : getLooters()) {
				Player player = creature.getOwner();
				if (player != null)
					player.sendPacket(new GroupOpenLotteryWindow(creature.getObjectId(), getCorpse().getInventory().getObjectId()));
			}
		}
		
		public synchronized void updatePreferences(CreatureObject looter, List<SWGObject> objects) {
			preferences.put(looter, objects);
			if (preferences.size() == getLooters().size())
				commitLottery();
		}
		
		public synchronized void commitLottery() {
			if (committed.getAndSet(true))
				return;
			// Build item map
			Map<SWGObject, List<CreatureObject>> itemPreferences = new HashMap<>();
			for (Entry<CreatureObject, List<SWGObject>> preference : preferences.entrySet()) {
				for (SWGObject item : preference.getValue())
					itemPreferences.computeIfAbsent(item, i -> new ArrayList<>()).add(preference.getKey());
			}
			
			// Notify the lottery is done
			for (CreatureObject creature : getLooters()) {
				Player player = creature.getOwner();
				if (player != null)
					player.sendPacket(new GroupCloseLotteryWindow(creature.getObjectId(), getCorpse().getInventory().getObjectId()));
			}
			
			// Give out items
			Random random = new Random();
			for (Entry<SWGObject, List<CreatureObject>> itemEntry : itemPreferences.entrySet()) {
				List<CreatureObject> looters = itemEntry.getValue();
				assert !looters.isEmpty();
				if (itemEntry.getKey() instanceof CreditObject && ((CreditObject) itemEntry.getKey()).getAmount() >= looters.size()) {
					splitCredits((CreditObject) itemEntry.getKey(), itemEntry.getValue());
				} else {
					CreatureObject winner = looters.get(random.nextInt(looters.size()));
					transferItem(winner, itemEntry.getKey());
				}
			}
		}
		
	}
	
}

File: src/main/java/com/projectswg/holocore/services/gameplay/combat/loot/GrantLootService.kt
/***********************************************************************************
 * Copyright (c) 2025 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is an emulation project for Star Wars Galaxies founded on            *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create one or more emulators which will provide servers for      *
 * players to continue playing a game similar to the one they used to play.        *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.services.gameplay.combat.loot

import com.projectswg.common.data.encodables.oob.ProsePackage
import com.projectswg.common.data.encodables.oob.StringId
import com.projectswg.common.network.packets.swg.zone.PlayMusicMessage
import com.projectswg.common.network.packets.swg.zone.StopClientEffectObjectByLabelMessage
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupCloseLotteryWindow
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupOpenLotteryWindow
import com.projectswg.common.network.packets.swg.zone.object_controller.loot.GroupRequestLotteryItems
import com.projectswg.holocore.intents.gameplay.combat.*
import com.projectswg.holocore.intents.support.global.chat.SystemMessageIntent
import com.projectswg.holocore.intents.support.global.chat.SystemMessageIntent.Companion.broadcastPersonal
import com.projectswg.holocore.intents.support.global.network.InboundPacketIntent
import com.projectswg.holocore.intents.support.objects.DestroyObjectIntent
import com.projectswg.holocore.intents.support.objects.OpenContainerIntent
import com.projectswg.holocore.resources.gameplay.combat.loot.LootType
import com.projectswg.holocore.resources.support.data.server_info.mongodb.PswgDatabase.config
import com.projectswg.holocore.resources.support.objects.permissions.ContainerResult
import com.projectswg.holocore.resources.support.objects.permissions.ReadWritePermissions
import com.projectswg.holocore.resources.support.objects.swg.SWGObject
import com.projectswg.holocore.resources.support.objects.swg.creature.CreatureObject
import com.projectswg.holocore.resources.support.objects.swg.custom.AIObject
import com.projectswg.holocore.resources.support.objects.swg.group.GroupObject
import com.projectswg.holocore.resources.support.objects.swg.group.LootRule
import com.projectswg.holocore.resources.support.objects.swg.tangible.CreditObject
import com.projectswg.holocore.services.support.objects.ObjectStorageService.ObjectLookup
import com.projectswg.holocore.utilities.HolocoreCoroutine
import com.projectswg.holocore.utilities.cancelAndWait
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import me.joshlarson.jlcommon.control.IntentHandler
import me.joshlarson.jlcommon.control.Service
import me.joshlarson.jlcommon.utilities.Arguments
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.stream.Collectors

class GrantLootService : Service() {
	private val lootRestrictions: MutableMap<CreatureObject, CorpseLootRestrictions> = ConcurrentHashMap()
	private val coroutineScope = HolocoreCoroutine.childScope()
	private val lootRange: Int = config.getInt(this, "lootRange", 16)

	override fun stop(): Boolean {
		coroutineScope.cancelAndWait()
		return super.stop()
	}

	@IntentHandler
	private fun handleLootGeneratedIntent(lgi: LootGeneratedIntent) {
		val corpse = lgi.corpse

		val corpseWorldLocation = corpse.worldLocation
		val mostHated = corpse.mostHated
		if (mostHated == null || !mostHated.isPlayer) return
		val groupId = mostHated.groupId

		val looters = ArrayList<CreatureObject>()
		if (groupId != 0L) {
			val group = checkNotNull(ObjectLookup.getObjectById(groupId) as GroupObject?)
			val masterLooter = group.lootMaster
			when (group.lootRule) {
				LootRule.MASTER_LOOTER                                   -> if (masterLooter != null) looters.add(masterLooter)
				LootRule.FREE_FOR_ALL, LootRule.RANDOM, LootRule.LOTTERY -> {
					// Get all looters within range
					looters.addAll(group.groupMemberObjects.stream().filter { it.worldLocation.flatDistanceTo(corpseWorldLocation) <= lootRange }.toList())
				}
			}
			if (group.lootRule == LootRule.LOTTERY) {
				lootRestrictions[corpse] = LotteryLootRestrictions(corpse, looters)
				corpse.inventory.containerPermissions = ReadWritePermissions.from(looters)
				return
			} else if (group.lootRule == LootRule.RANDOM) {
				lootRestrictions[corpse] = RandomLootRestrictions(corpse, looters)
				corpse.inventory.containerPermissions = ReadWritePermissions.from(looters)
				return
			}
		} else {
			looters.add(mostHated)
		}
		lootRestrictions[corpse] = StandardLootRestrictions(corpse, looters)
		val inventory = corpse.inventory
		inventory.containerPermissions = ReadWritePermissions.from(looters)
		for (loot in inventory.containedObjects) loot.containerPermissions = ReadWritePermissions.from(looters)
	}

	@IntentHandler
	private fun handleDestroyObjectIntent(doi: DestroyObjectIntent) {
		// For when the corpse is gone
		val obj = doi.obj
		if (obj is CreatureObject) {
			val restrictions = lootRestrictions.remove(obj)
			restrictions?.setValid(false)
		}
	}

	@IntentHandler
	private fun handleInboundPacketIntent(ipi: InboundPacketIntent) {
		val request = ipi.packet as? GroupRequestLotteryItems ?: return
		val player = ipi.player
		val requestObject = ObjectLookup.getObjectById(request.inventoryId)?.parent ?: return
		if (requestObject !is CreatureObject) return
		val restriction = lootRestrictions[requestObject] as? LotteryLootRestrictions ?: return
		restriction.updatePreferences(player.creatureObject, request.requestedItems.stream().map<SWGObject> { ObjectLookup.getObjectById(it) }.filter(Objects::nonNull).collect(Collectors.toList()))
	}

	@IntentHandler
	private fun handleLootRequestIntent(lri: LootRequestIntent) {
		val player = lri.player
		val looter = player.creatureObject
		val corpse = lri.target

		Arguments.validate(corpse is AIObject, "Attempted to loot a non-AI object")
		val restrictions = lootRestrictions[corpse] ?: // No loot was generated. Do nothing.
		return
		if (!restrictions.canLoot(looter)) {
			broadcastPersonal(player, "You don't have permission to loot '" + corpse.objectName + '\'')
			return
		}

		if (restrictions is LotteryLootRestrictions) {
			if (restrictions.isStarted()) return  // Don't keep re-starting the lottery

			coroutineScope.launch {
				// Close the windows after 30 seconds, and grant the items won
				delay(30_000L)
				if (restrictions.isValid()) {
					restrictions.commitLottery()
				}
			}
		}
		restrictions.handle(looter, lri.type)
	}

	@IntentHandler
	private fun handleLootItemIntent(lii: LootItemIntent) {
		val looter = lii.looter
		val corpse = lii.corpse
		val item = lii.item

		Arguments.validate(corpse is AIObject, "Attempted to loot a non-AI object")
		val restrictions = lootRestrictions[corpse] ?: // No loot was generated. Do nothing.
		return
		if (!restrictions.canLoot(looter)) {
			broadcastPersonal(looter.owner!!, "You don't have permission to loot '" + corpse.objectName + '\'')
			return
		}

		restrictions.loot(looter, item)
	}

	private abstract class CorpseLootRestrictions(val corpse: CreatureObject, val looters: List<CreatureObject>) {
		private val valid = AtomicBoolean(true)

		fun isValid(): Boolean {
			return valid.get()
		}

		fun canLoot(looter: CreatureObject): Boolean {
			return looters.contains(looter)
		}

		fun setValid(valid: Boolean) {
			this.valid.set(valid)
		}

		@Synchronized
		open fun loot(looter: CreatureObject, item: SWGObject) {
			throw UnsupportedOperationException("this loot restriction cannot do manual looting")
		}

		abstract fun handle(looter: CreatureObject, type: LootType)

		@Synchronized
		protected fun transferItem(looter: CreatureObject, item: SWGObject) {
			if (!corpse.inventory.containedObjects.contains(item) || !canLoot(looter)) {
				return
			}
			val player = checkNotNull(looter.owner)
			when (item.moveToContainer(looter, looter.inventory)) {
				ContainerResult.SUCCESS        -> {
					var itemName = item.objectName
					if (itemName.isEmpty()) itemName = item.stringId.toString()

					if (item is CreditObject) {
						onLootedCredits(looter, item.amount)
					} else {
						SystemMessageIntent(player, ProsePackage("StringId", StringId("spam", "loot_item_self"), "TU", itemName, "TT", corpse.objectName)).broadcast()
					}
					onLooted(looter, corpse)
				}

				ContainerResult.CONTAINER_FULL -> {
					SystemMessageIntent(player, "@container_error_message:container03").broadcast()
					player.sendPacket(PlayMusicMessage(0, "sound/ui_danger_message.snd", 1, false))
				}

				ContainerResult.NO_PERMISSION  -> {
					SystemMessageIntent(player, "@container_error_message:container08").broadcast()
					player.sendPacket(PlayMusicMessage(0, "sound/ui_negative.snd", 1, false))
				}

				ContainerResult.SLOT_NO_EXIST  -> {
					SystemMessageIntent(player, "@container_error_message:container06").broadcast()
					player.sendPacket(PlayMusicMessage(0, "sound/ui_negative.snd", 1, false))
				}

				ContainerResult.SLOT_OCCUPIED  -> {
					SystemMessageIntent(player, "@container_error_message:container04").broadcast()
					player.sendPacket(PlayMusicMessage(0, "sound/ui_negative.snd", 1, false))
				}
			}
		}

		protected fun splitCredits(target: CreditObject, looters: Collection<CreatureObject>) {
			val amount = target.amount / looters.size
			assert(amount > 0)

			for (looter in looters) {
				looter.addToBank(amount)
				onLootedCredits(looter, amount)
			}

			DestroyObjectIntent(target).broadcast()
		}

		protected fun onLooted(looter: CreatureObject, corpse: CreatureObject) {
			val lootInventory = corpse.inventory
			if (lootInventory.containedObjects.isEmpty()) {
				CorpseLootedIntent(corpse).broadcast()

				val looterGroupId = looter.groupId
				if (looterGroupId != 0L) {
					val killerGroup = checkNotNull(ObjectLookup.getObjectById(looterGroupId) as GroupObject?)
					for (groupMember in killerGroup.groupMemberObjects) groupMember.owner!!.sendPacket(StopClientEffectObjectByLabelMessage(corpse.objectId, "lootMe", false))
				} else {
					looter.owner!!.sendPacket(StopClientEffectObjectByLabelMessage(corpse.objectId, "lootMe", false))
				}
			}
		}

		protected fun onLootedCredits(looter: CreatureObject, amount: Long) {
			// Perhaps "prose_coin_loot_no_target" is the proper string?
			SystemMessageIntent(looter.owner!!, ProsePackage("StringId", StringId("base_player", "prose_transfer_success"), "DI", amount.toInt())).broadcast()
		}
	}

	private class StandardLootRestrictions(corpse: CreatureObject, looters: List<CreatureObject>) : CorpseLootRestrictions(corpse, looters) {
		@Synchronized
		override fun loot(looter: CreatureObject, item: SWGObject) {
			transferItem(looter, item)
		}

		@Synchronized
		override fun handle(looter: CreatureObject, type: LootType) {
			val lootInventory = corpse.inventory
			val lootItems = lootInventory.containedObjects
			when (type) {
				LootType.LOOT     -> if (!lootItems.isEmpty()) OpenContainerIntent(looter, lootInventory, "").broadcast()
				LootType.LOOT_ALL -> for (loot in lootItems) loot(looter, loot)
			}
		}
	}

	private class RandomLootRestrictions(corpse: CreatureObject, looters: List<CreatureObject>) : CorpseLootRestrictions(corpse, looters) {
		@Synchronized
		override fun loot(looter: CreatureObject, item: SWGObject) {
			if (item is CreditObject && item.amount >= looters.size) splitCredits(item, looters)
			else transferItem(looter, item)
		}

		@Synchronized
		override fun handle(looter: CreatureObject, type: LootType) {
			val lootItems = corpse.inventory.containedObjects
			for (loot in lootItems) loot(looter, loot)
		}
	}

	private class LotteryLootRestrictions(corpse: CreatureObject, looters: List<CreatureObject>) : CorpseLootRestrictions(corpse, looters) {
		private val preferences: MutableMap<CreatureObject, List<SWGObject>> = HashMap()
		private val started = AtomicBoolean(false)
		private val committed = AtomicBoolean(false)

		fun isStarted(): Boolean {
			return started.get()
		}

		@Synchronized
		override fun handle(looter: CreatureObject, type: LootType) {
			if (started.getAndSet(true)) return
			if (corpse.inventory.containedObjects.isEmpty()) return
			LootLotteryStartedIntent(corpse).broadcast()
			for (creature in looters) {
				val player = creature.owner
				player?.sendPacket(GroupOpenLotteryWindow(creature.objectId, corpse.inventory.objectId))
			}
		}

		@Synchronized
		fun updatePreferences(looter: CreatureObject, objects: List<SWGObject>) {
			preferences[looter] = objects
			if (preferences.size == looters.size) commitLottery()
		}

		@Synchronized
		fun commitLottery() {
			if (committed.getAndSet(true)) return
			// Build item map
			val itemPreferences: MutableMap<SWGObject, MutableList<CreatureObject>> = HashMap()
			for ((key, value) in preferences) {
				for (item in value) itemPreferences.computeIfAbsent(item) { ArrayList() }.add(key)
			}


			// Notify the lottery is done
			for (creature in looters) {
				val player = creature.owner
				player?.sendPacket(GroupCloseLotteryWindow(creature.objectId, corpse.inventory.objectId))
			}


			// Give out items
			val random = Random()
			for ((key, looters) in itemPreferences) {
				assert(looters.isNotEmpty())
				if (key is CreditObject && key.amount >= looters.size) {
					splitCredits(key, looters)
				} else {
					val winner = looters[random.nextInt(looters.size)]
					transferItem(winner, key)
				}
			}
		}
	}
}